### DISCUSSION TITLE
Option to configure web proxy

### DISCUSSION THREAD

**Discussion by Fx4-X** [directus/directus#6386] [Archive]
Hello,

Do we have an existing option to set web proxy (client proxy) for communication out of Directus ?
Communication to web hook, OAuth, custom api hook.

Edit : scenario
The Directus is installed in a local network, only way to reach internet is via the proxy.
The communication originated (ex : oAuth, webhook...) from Directus also has to be go via the proxy for internet trafic, but at the moment there are no option to set the proxy config.

**Comment by philleepflorence** [comment]
See if these help with communication out of Directus:
- [Webhooks Guide: Directus Docs](https://docs.directus.io/guides/webhooks/)
- [Custom API Hooks Guide: Directus Docs](https://docs.directus.io/guides/api-hooks/)

**Reply by Fx4-X** [reply]
It's not about webhook or custom api hook, it's to enable such communication to go via proxy.

sorry if I confused you :-)

**Reply by philleepflorence** [reply]
Ah sorry! Thanks for the clarification.

How close is `PUBLIC_URL` in [Configuring the API - oAuth Guide: Directus Docs](https://docs.directus.io/guides/api-config/#oauth-single-sign-on-sso-openid) related to your question?

**Comment by rijkvanzanten** [comment]
There's no proxy configuration currently 👍🏻

**Reply by Fx4-X** [reply]
Thanks @rijkvanzanten  for the clarification. 

Do you know if there is any workaround possible ? or any patch ?

**Reply by rijkvanzanten** [reply]
No idea! Never really had to rely on a proxy in this fashion, so I don't know. From what I can tell from some googling, it looks like we'll have to add proxying logic to every http call that's made, so that will take some reworking of those system data flows.

**Reply by Fx4-X** [reply]
Currently I need to use proxy for oAuth,
I found this https://github.com/simov/grant/blob/058665b2734538f64f265e0434b3bb39adf39e7f/examples/request-options/proxy-agent.js#L5

looks like I could add a patch in Directus, but struggling to figure out in which file.
Could you please help me ?

**Reply by rijkvanzanten** [reply]
We'll have to find every instance where outgoing requests are made, and then add functionality to allow for proxying to all those cases. I've never looked into supporting proxying, so I'm not quite sure what's required to make that happen.

As for grant's config specifically, that's here: https://github.com/directus/directus/blob/main/api/src/grant.ts

**Reply by Fx4-X** [reply]
Thanks @rijkvanzanten for the config file.

@simov : sorry for dragging you to this,
I need help to add http proxy config for grant requests made out of Directus.
Could you please help with this magic config :-)

**Reply by simov** [reply]
Hi @Fx4-X, the file that @rijkvanzanten, pointed you to https://github.com/directus/directus/blob/main/api/src/grant.ts contains the Grant configuration. However, Grant supports extended configuration https://github.com/simov/grant/blob/master/grant.d.ts#L10-L30 where the configuration from the above mentioned files is only the contents of the `config` key. That's how you can have additional configuration outside of the regular one, such as the `request` one, which can contain any Node.js core HTTP option such as `agent` and any `request-compose` option which is the underlying HTTP client being used in Grant.

So if Directus supports proxying through some environment variable for example, then you can configure your Grant instance inside Directus to use that, though you will need additional module for that as mentioned here https://github.com/simov/grant/blob/058665b2734538f64f265e0434b3bb39adf39e7f/examples/request-options/proxy-agent.js#L5. These are the examples from the underlying HTTP client: https://github.com/simov/request-compose/blob/master/examples/misc-proxy-agent.js and https://github.com/simov/request-compose/blob/master/examples/misc-tunnel-agent.js

**Reply by rijkvanzanten** [reply]
(unrelated: @simov you should look into setting up GH sponsors 🙂 would be nice to send you a little thank you for all your help)

**Reply by Fx4-X** [reply]
Thanks @simov and @rijkvanzanten for the support,

I installed the 'proxy-agent' npm package
and added the proxy config in directus/node_modules/directus/dist/grant.js file (Running Directus in Docker)
I am not sure if I done it correctly :-)

The Directus grant.js config file looks like below: (3 lines added)

```js
"use strict";
/**
 * Grant is the oAuth library
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const env_1 = __importDefault(require("./env"));
const to_array_1 = require("./utils/to-array");
const get_config_from_env_1 = require("./utils/get-config-from-env");
const enabledProviders = to_array_1.toArray(env_1.default.OAUTH_PROVIDERS).map((provider) => provider.toLowerCase());

var proxy = require('proxy-agent')                                           //Added line
var agent = new proxy('http://proxyserver.com:8090')            //Added line

const config = {
    defaults: {
        origin: env_1.default.PUBLIC_URL,
        transport: 'session',
        prefix: '/auth/oauth',
        response: ['tokens', 'profile'],
        request: {agent},                                                                //Added line
    },
};
for (const provider of enabledProviders) {
    config[provider] = get_config_from_env_1.getConfigFromEnv(`OAUTH_${provider.toUpperCase()}_`, undefined, 'underscore');
}
exports.default = config;
```

Still no luck, I outputted the hookPayload (https://github.com/directus/directus/blob/44ffd60174f47df9c8a7eab7c2673e2bca0b43bc/api/src/controllers/auth.ts#L287)

and this what it read
```
{
  error: 'request-compose: timeout',
  profile: { error: 'request-compose: timeout' }
}
```

I inspected the network traffic and its still not using the proxy.
could that be grant.js config file still missing something.

**Reply by simov** [reply]
@Fx4-X your configuration object have to look like this:

```js
const config = {
  config: {defaults: {...}},
  request: {agent}
}
```

and then the rest of the changes.

**Reply by Fx4-X** [reply]
The grant.js config looks like below :
```js
const config = {
  config: {    defaults: {
        origin: env_1.default.PUBLIC_URL,
        transport: 'session',
        prefix: '/auth/oauth',
        response: ['tokens', 'profile'],
    },
  },
  request: {agent}
}
```

its generate this error "Grant: missing or misconfigured provider"

**Reply by Fx4-X** [reply]
@simov , how about adding a patch in grant/lib/client.js
not sure if this the right way or file, 
but I found the "request-compose" and "user-agent" config, and thought there could be a option to append proxy here.

is that could be plan B ? :-)

**Reply by simov** [reply]
No, Grant already supports that through the `request` configuration. It is an option that you can configure, it is not built into the library.

**Reply by rijkvanzanten** [reply]
In any case, I would strongly suggest not patching any code under `node_modules` directly in your own deployment, as that's bound to cause issues once the package has an update, or when you are forced to reinstall the node modules for other reasons  (like node upgrades, re-deploys in a different environment, CI testing, etc etc)

**Comment by Fx4-X** [comment]
I guess the Directus expect only the "defaults" and not the "config.defaults" at https://github.com/directus/directus/blob/18d6c0edbe4ccb35d4bd45fc6f5cde988e8e07c3/api/src/controllers/auth.ts#L267

hence below config format not working.
```js
const config = {
  config: {defaults: {...}}, -> this not working, and I guess Directus designed to pass only defaults: {...} to grant.
  request: {agent}
}
```
and getting the error "Grant: missing or misconfigured provider"

@rijkvanzanten could you help or confirm if we could use the above config format and Directus still work.

**Reply by rijkvanzanten** [reply]
I can't guarantee that. Directus currently doesn't officially support proxies for oAuth / other external calls, so I can't guarantee that any workaround patches will work (and remain to work moving forwards)

**Reply by Fx4-X** [reply]
Thanks @rijkvanzanten & @simov for all the guidance,

at last its start working :-)

Steps to follow : (I am using docker)

1. Install Directus using [these steps](https://docs.directus.io/guides/installation/docker/)
2. Config proxy on npm inside the Docker
        `docker exec -i <container name> npm config set proxy http://proxy.server.com:8080`
3. Install proxy-agent
        `docker exec -i <container name> npm install proxy-agent`
4. remove proxy config from npm (its optional)
        `docker exec -i <container name> npm config rm proxy`
5. Go to the container app dir
        ```cd `docker inspect --format={{.GraphDriver.Data.MergedDir}} <container name>` ```
6. Edit Grant config in Directus app and change below lines (vi directus/node_modules/directus/dist/grant.js), [grant.ts](https://github.com/directus/directus/blob/main/api/src/grant.ts)
```js
var proxy = require('proxy-agent');                        // import newly installed npm package
var agent = new proxy('http://proxy.server.com:8080');     // declare proxy config

const config = {
    config : { defaults: {                                 // move "defaults" into the "config" to match with grant requirement.
        origin: env_1.default.PUBLIC_URL,
        transport: 'session',
        prefix: '/auth/oauth',
        response: ['tokens', 'profile']
    } },
    request: {agent, timeout: 30}                          // add proxy config to the grant
};
for (const provider of enabledProviders) {
    config.config[provider] = get_config_from_env_1.getConfigFromEnv(`OAUTH_${provider.toUpperCase()}_`, undefined, 'underscore');       // move the oAuth providers config to the "config"
}
exports.default = config;
```
7. Edit the auth.js file to compensate the changes in "defaults" (vi directus/node_modules/directus/dist/controllers/auth.js) 
    search for below line https://github.com/directus/directus/blob/5936c8b2d9135f9df9b42b4e572f6ac604c52d75/api/src/controllers/auth.ts#L227

and change it from 
```js
const config = { ...grant_2.default};
```
to 
```js
const config = { ...grant_2.default.config };
```
8. Restart the container and that it Directus oAuth/Grant should be able to use Proxy, enjoy :-)

I know this is not perfect but at least its working for me at the moment.
and need to repeat the steps at every upgrade, (could create a local patched docker image to make everything smoother)

**Reply by simov** [reply]
Thanks for the feedback @Fx4-X, I'm glad it worked out for you.

**Comment by NilsBaumgartner1994** [comment]
Please add this as a feature.
